arrow-libs/core/arrow-autoclose/src/commonMain/kotlin/arrow/AutoCloseScope.kt

```diff
@@ -63,43 +63,43 @@ import kotlin.coroutines.cancellation.CancellationException
  */
 public inline fun <A> autoCloseScope(block: AutoCloseScope.() -> A): A {
   val scope = DefaultAutoCloseScope()
+  var finalized = false
   return try {
     block(scope)
-      .also { scope.close(null) }
-  } catch (e: CancellationException) {
-    scope.close(e) ?: throw e
   } catch (e: Throwable) {
-    scope.close(e.throwIfFatal()) ?: throw e
+    finalized = true
+    scope.close(e.throwIfFatal())
+    throw e
+  } finally {
```

Comment:
Using `finally` to ensure closing even in non-local returns (or `break`s even in Kotlin 2.2)